cpp/src/arrow/compute/exec/sink_node.cc

```diff
@@ -232,6 +232,77 @@ class ConsumingSinkNode : public ExecNode {
   std::shared_ptr<SinkNodeConsumer> consumer_;
 };
 
+/**
+ * @brief This node is an extension on ConsumingSinkNode
+ * to facilitate to get the output from an execution plan
+ * as a table. We define a custom SinkNodeConsumer to
+ * enable this functionality.
+ */
+
+struct TableSinkNodeConsumer : public arrow::compute::SinkNodeConsumer {
+ public:
+  TableSinkNodeConsumer(std::shared_ptr<Table>* out,
+                        std::shared_ptr<Schema> output_schema, MemoryPool* pool,
+                        Future<> finish)
+      : out_(out),
+        output_schema_(output_schema),
+        pool_(pool),
+        finish_(std::move(finish)) {}
+
+  Status Consume(ExecBatch batch) override {
+    ARROW_ASSIGN_OR_RAISE(auto rb, batch.ToRecordBatch(output_schema_, pool_));
+    if (rb) {
+      batch_vector.push_back(rb);
+    } else {
+      return Status::Invalid("Invalid ExecBatch consumed");
+    }
+    return Status::OK();
+  }
+
+  Future<> Finish() override {
+    ARROW_ASSIGN_OR_RAISE(auto table, Table::FromRecordBatches(batch_vector));
+    *out_ = table;
```

Comment:
nit, but this can just be `ARROW_ASSIGN_OR_RAISE(*out_, Table::FromRecordBatches(std::move(batch_vector)));`

ARROW_ASSIGN_OR_RAISE effectively lets you put anything on the LHS, there is no need to declare a variable there unless you want to.